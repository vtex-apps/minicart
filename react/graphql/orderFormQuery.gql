query MinicartQuery {
  orderForm {
    cacheId
    orderFormId
    value
    totalizers {
      id
      name
      value
    }
    items {
      id
      name
      imageUrl
      detailUrl
      skuName
      quantity
      sellingPrice
      listPrice
      parentItemIndex
      parentAssemblyBinding
      cartIndex
      assemblyOptions {
        added {
          item {
            name
            sellingPrice
            quantity
          }
          normalizedQuantity
          choiceType
          extraQuantity
        }
        removed {
          removedQuantity
          initialQuantity
          name
        }
        parentPrice
      }
      seller
    }
    shippingData {
      address {
        id
        neighborhood
        complement
        number
        street
        postalCode
        city
        reference
        addressName
        addressType
        geoCoordinates
        state
        receiverName
        country
      }
      availableAddresses {
        id
        neighborhood
        complement
        number
        street
        postalCode
        city
        reference
        addressName
        addressType
        geoCoordinates
        state
        receiverName
        country
      }
    }
    clientProfileData {
      email
      firstName
    }
    storePreferencesData {
      countryCode
      currencyCode
      timeZone
    }
    checkedInPickupPointId
    isCheckedIn
  }
  minicart @client {
    items
    orderForm
    isOpen
  }
}
