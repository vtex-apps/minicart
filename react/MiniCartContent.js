import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql } from 'react-apollo'
import { injectIntl, intlShape } from 'react-intl'
import updateItemsMutation from './graphql/updateItemsMutation.gql'
import orderFormQuery from './graphql/orderFormQuery.gql'
import MiniCartItem from './MiniCartItem'
import Button from '@vtex/styleguide/lib/Button'
import { Price } from '@vtex/product-details'
import Spinner from '@vtex/styleguide/lib/Spinner'
import { MiniCartPropTypes } from './MiniCartPropTypes'

import './global.css'

/**
 * Minicart content component
 */
class MiniCartContent extends Component {
  static propTypes = {
    /* Products in the cart */
    data: MiniCartPropTypes.data,
    /* Mutate function */
    mutate: PropTypes.func.isRequired,
    /* Label to appear when the minicart is empty */
    labelMiniCartEmpty: PropTypes.string,
    /* Label to appear in the finish shopping button */
    labelButton: PropTypes.string,
    /* Show remove item button or not */
    showRemoveButton: PropTypes.bool,
    onUpdateItemsQuantity: PropTypes.func,
    /* Internationalization */
    intl: intlShape.isRequired,
  }

  handleClickButton = () => location.assign('/checkout/#/cart')

  onRemoveItem = id => {
    const { mutate, data: { orderForm } } = this.props
    const itemPayload = orderForm.items.find(item => item.id === id)
    const index = orderForm.items.indexOf(itemPayload)
    const updatedItem = [itemPayload].map(item => {
      return {
        id: parseInt(item.id),
        index: index,
        quantity: 0,
        seller: 1,
      }
    })
    mutate({
      variables: {
        orderFormId: orderForm.orderFormId,
        items: updatedItem,
      },
      refetchQueries: [{ query: orderFormQuery }],
    }).then(() => {
      this.props.onUpdateItemsQuantity(this.props.data.orderForm.items.length - 1)
    })
  }

  renderWithoutItems = label => (
    <div className="vtex-minicart__item pa4 flex items-center justify-center relative bg-white">
      <span className="f5">{label}</span>
    </div>
  )

  renderMiniCartWithItems = (orderForm, label, showRemoveButton) => (
    <div className="flex flex-column relative" >
      <div className="bg-white">
        <div className="vtex-minicart__content pr4 pl4 overflow-auto">
          {orderForm.items.map(item => (
            <div className="flex flex-row" key={item.id}>
              <MiniCartItem {...item} removeItem={this.onRemoveItem} showRemoveButton={showRemoveButton} />
            </div>
          ))}
        </div>
        <div className="fl pa4">
          <Button primary onClick={this.handleClickButton}>{label}</Button>
        </div>
        <div className="fr pt4 mt2 mr4">
          <Price
            sellingPrice={orderForm.value}
            listPrice={orderForm.value}
            showLabels={false}
            showListPrice={false} />
        </div>
      </div>
    </div >
  )

  renderLoading = () => (
    <div className="vtex-minicart__item pa4 flex items-center justify-center relative bg-white">
      <Spinner />
    </div>
  )

  componentDidMount() {
    this.props.data.refetch().then(() => {
      this.props.onUpdateItemsQuantity(this.props.data.orderForm.items.length)
    })
  }

  render() {
    const { data, labelMiniCartEmpty, labelButton, intl, showRemoveButton } = this.props
    let content
    if (data.loading) {
      content = this.renderLoading()
    } else if (!data.orderForm || !data.orderForm.items.length) {
      const label = labelMiniCartEmpty || intl.formatMessage({ id: 'minicart-empty' })
      content = this.renderWithoutItems(label)
    } else {
      const label = labelButton || intl.formatMessage({ id: 'finish-shopping-button-label' })
      content = this.renderMiniCartWithItems(data.orderForm, label, showRemoveButton)
    }
    return content
  }
}

export default injectIntl(graphql(updateItemsMutation)(MiniCartContent))
