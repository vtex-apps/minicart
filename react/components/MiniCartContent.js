import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql } from 'react-apollo'
import { injectIntl, intlShape } from 'react-intl'
import { reduceBy, values } from 'ramda'

import Button from '@vtex/styleguide/lib/Button'
import Spinner from '@vtex/styleguide/lib/Spinner'
import ProductPrice from 'vtex.store-components/ProductPrice'

import updateItemsMutation from '../graphql/updateItemsMutation.gql'
import orderFormQuery from '../graphql/orderFormQuery.gql'
import MiniCartItem from './MiniCartItem'
import { MiniCartPropTypes } from '../propTypes'

import '../global.css'

/**
 * Minicart content component
 */
class MiniCartContent extends Component {
  static propTypes = {
    /* Function to be called when an item is removed */
    onUpdateItemsQuantity: PropTypes.func,
    /* Mutate function */
    mutate: PropTypes.func.isRequired,
    /* Internationalization */
    intl: intlShape.isRequired,
    /* Reused props */
    data: MiniCartPropTypes.data,
    labelMiniCartEmpty: MiniCartPropTypes.labelMiniCartEmpty,
    labelButton: MiniCartPropTypes.labelButtonFinishShopping,
    showRemoveButton: MiniCartPropTypes.showRemoveButton,
    enableQuantitySelector: MiniCartPropTypes.enableQuantitySelector,
    maxQuantity: MiniCartPropTypes.maxQuantity,
  }

  constructor(props) {
    super(props)
    this.state = { showSpinner: false }
  }

  handleClickButton = () => location.assign('/checkout/#/cart')

  onRemoveItem = id => {
    const { mutate, data: { orderForm } } = this.props
    const itemPayload = orderForm.items.find(item => item.id === id)
    const index = orderForm.items.indexOf(itemPayload)
    const updatedItem = [itemPayload].map(item => {
      return {
        id: parseInt(item.id),
        index: index,
        quantity: 0,
        seller: 1,
      }
    })
    mutate({
      variables: {
        orderFormId: orderForm.orderFormId,
        items: updatedItem,
      },
      refetchQueries: [{ query: orderFormQuery }],
    }).then(() => {
      this.props.onUpdateItemsQuantity(this.props.data.orderForm.items.length - 1)
    })
  }

  onUpdateItems = (id, quantity) => {
    this.setState({ showSpinner: true })
    const { mutate, data: { orderForm } } = this.props
    const itemPayload = orderForm.items.find(item => item.id === id)
    const index = orderForm.items.indexOf(itemPayload)
    const updatedItem = [itemPayload].map(item => {
      return {
        id: parseInt(item.id),
        index: index,
        quantity: quantity,
        seller: 1,
      }
    })
    mutate({
      variables: {
        orderFormId: orderForm.orderFormId,
        items: updatedItem,
      },
      refetchQueries: [{ query: orderFormQuery }],
    }).then(() => {
      this.setState({
        showSpinner: false,
      })
    })
  }

  renderWithoutItems = label => (
    <div className="vtex-minicart__item pa4 flex items-center justify-center relative bg-white">
      <span className="f5">{label}</span>
    </div>
  )

  renderMiniCartWithItems = (
    orderForm,
    label,
    showRemoveButton,
    enableQuantitySelector,
    maxQuantity,
    showSpinner
  ) => {
    const items = values(
      reduceBy(
        (acc, item) =>
          acc ? { ...acc, quantity: acc.quantity + item.quantity } : item,
        undefined,
        item => item.id,
        orderForm.items
      )
    )
    return (
      <div className="flex flex-column relative" >
        <div className="bg-white">
          <div className="vtex-minicart__content pr4 pl4 overflow-auto overflow-x-hidden">
            {items.map(item => (
              <MiniCartItem
                {...item}
                key={item.id}
                removeItem={this.onRemoveItem}
                updateItem={this.onUpdateItems}
                showRemoveButton={showRemoveButton}
                enableQuantitySelector={enableQuantitySelector}
                maxQuantity={maxQuantity}
              />
            ))}
          </div>
          <div className="fl pa4">
            <Button primary onClick={this.handleClickButton}>{label}</Button>
          </div>
          <div className="flex flex-row fr pt4 mt2 mr4">
            {showSpinner && <Spinner size={18} />}
            <ProductPrice
              sellingPrice={orderForm.value}
              listPrice={orderForm.value}
              showLabels={false}
              showListPrice={false}
            />
          </div>
        </div>
      </div>
    )
  }

  renderLoading = () => (
    <div className="vtex-minicart__item pa4 flex items-center justify-center relative bg-white">
      <Spinner />
    </div>
  )

  componentDidMount() {
    this.props.data.refetch().then(() => {
      this.props.onUpdateItemsQuantity(this.props.data.orderForm.items.length)
    })
  }

  render() {
    const {
      data,
      labelMiniCartEmpty,
      labelButton,
      intl,
      showRemoveButton,
      enableQuantitySelector,
      maxQuantity,
    } = this.props
    const { showSpinner } = this.state

    if (data.loading) {
      return this.renderLoading()
    }

    if (!data.orderForm || !data.orderForm.items.length) {
      const label = labelMiniCartEmpty || intl.formatMessage({ id: 'minicart-empty' })
      return this.renderWithoutItems(label)
    }

    const label = labelButton || intl.formatMessage({ id: 'finish-shopping-button-label' })

    return this.renderMiniCartWithItems(
      data.orderForm,
      label,
      showRemoveButton,
      enableQuantitySelector,
      maxQuantity,
      showSpinner
    )
  }
}

export default injectIntl(graphql(updateItemsMutation)(MiniCartContent))
