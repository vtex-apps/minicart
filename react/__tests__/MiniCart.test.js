/* eslint-env jest */
import React from 'react'
import { render } from 'react-testing-library'
import { MockedProvider } from 'react-apollo/test-utils'
import orderFormQuery from '../graphql/orderFormQuery.gql'
import { IntlProvider } from 'react-intl'

import MiniCartContent from '../MiniCartContent'

describe('MiniCart component', () => {
  let minicartWithProducts

  beforeEach((done) => {
    const messages = require('../locales/en-US')
    const mockedOrderFrom = {
      orderFormId: '1092',
      items: [
        {
          id: '31',
          detailUrl: 'celular',
          imageUrl: 'https://raw.githubusercontent.com/vtex-apps/product-summary/feature/product-image/images/500x500-img-pro20.png',
          name: 'Celular',
          quantity: 1,
          sellingPrice: 154000,
          listPrice: 1000,
          skuName: 'Moto X4',
        }
      ],
      value: 1632000,
    }

    minicartWithProducts = render(
      <MockedProvider
        mocks={[
          {
            request: { query: orderFormQuery },
            result: { data: { orderForm: mockedOrderFrom } },
          },
        ]}
      >
        <IntlProvider
          locale="en-US"
          messages={messages}
        >
          <MiniCartContent />
        </IntlProvider>
      </MockedProvider>
    )

    setTimeout(() => {
      done()
    }, 0)
  })

  it('should be rendered', () => {
    expect(minicartWithProducts).toBeDefined()
  })

  it('should match snapshot', () => {
    expect(minicartWithProducts).toMatchSnapshot()
  })

  it('minicart with products should render three minicart items', () => {
    expect(minicartWithProducts.container.querySelectorAll('.vtex-minicart__item').length).toBe(1)
  })
})
