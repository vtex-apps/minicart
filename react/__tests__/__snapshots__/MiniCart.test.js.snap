// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniCart component should match snapshot 1`] = `
Object {
  "container": <div>
    <div
      class="vtex-minicart__item pa4 shadow-4 flex items-center justify-center"
    >
      <span
        class="f5"
      >
        Your bag is empty! :(
      </span>
    </div>
  </div>,
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
